### PR TITLE
Build phar with box and bump dependency versions

### PR DESCRIPTION
**Pull request by jakzal**
re https://github.com/pdepend/pdepend/issues/371 https://github.com/pdepend/pdepend/issues/364

Advantages of dropping legacy PHP versions:

* less maintenance burden

Advantages of box:

* standardised way to build phar files
* it's faster
* possibility to scope phar classes (not part of this PR)

Disadvantages of box:

* it diverges from the ant build system set up for this project. It only makes sense if everything else is migrated over and the ant setup is removed (it seems to be unmaintained).

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -50,3 +32,15 @@ after_script:
 
 notifications:
     irc: "irc.freenode.org#pdepend"
+
+deploy:
+    provider: releases
+    api_key:
+        secure: LLotlFFZYAud5hxpxitvReMFanKIU5r6gtFsuxstbcr0tYQSxK8Wr0Uo0OPHqBT6c3LqZivqSIFkxEl7Hg/voutHpNDV5jCnLUmvkyFGSp4eNBy+YNoabGR8Hg3Y2a+DfD+fJmFisbMfbfY9sulq24GghTfToVX75vhW5fsTZWGTyvDbl0A5EPxNhIQzrCVfJfWT7f0TN0T/+hnGzoqEaofKfJctkJpW7R4e+DQ0s7J1ydG1cEBGC7n3PdvHp/Ab5EqnBrpvjIrH4ljsVl+fAEAJjN7qjEQNQp4M2xeb0DQ5EaPXi/NukPq/0itb5IgHg2F+RHdCYDoWeP16hl/dkgB9aZacnr9aiPDL0s2kNJ8ujmReyFqcw50YYcUmlZT61VeZ0qGiVmwRQ+eTdf3il9WosAaJ349z2aRrr4aR4CvAoSxmpThrzU0TwIKnmey1Rb+4kCWsJTwEg46+tZWuQgOpdv/la9SCSqhWbB6Qn8KqYlm1eK1hpxKS+KxVeby1k7lloO02Y37tjEpyt/2u0z2F9KNojChTxUeSzW66cEi4Ox6q+y3zi++6qj0BVzxbmjxuFSzhe9ZaL3gyWmOKuSywVnJ5UmQw9CL4gg91iZWNbKRcpO5H1oiBSdK5jMFcfCof1UmAuL34ygFbNVk7Aco6ssgBUROqTEjjW2acrzQ=
```
Comment (jakzal) [inline .travis.yml]: This key will need to be updated for pdepend/pdepend. See `travis setup releases`.

File: .travis.yml [hunk 2]
```diff
@@ -50,3 +32,15 @@ after_script:
...
+    file:
+        - build/pdepend.phar
+    skip_cleanup: true
+    on:
+        tags: true
+        repo: jakzal/pdepend
```
Comment (jakzal) [inline .travis.yml]: This will need to be updated to `pdepend/pdepend`.

File: phpunit.xml.dist [hunk 3]
```diff
@@ -1,5 +1,7 @@
 <?xml version="1.0"?>
-<phpunit bootstrap="src/test/php/PDepend/bootstrap.php">
+<phpunit bootstrap="src/test/php/PDepend/bootstrap.php"
+         beStrictAboutTestsThatDoNotTestAnything="false"
```
Comment (jakzal) [inline phpunit.xml.dist]: There's loads of tests with no assertions, but one thing at a time.

**Comment by emirb** [comment]
cc @manuelpichler @ravage84

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,58 @@
+default: build
+
+build: install test
+.PHONY: build
```
Comment (theofidry) [inline Makefile]: I think it's more standard to put the phone _before_, but that's just CS anyway...

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+vendor: install
+
+vendor/bin/phpunit: install
+
+tools/phpunit: vendor/bin/phpunit
+	ln -sf ../vendor/bin/phpunit tools/phpunit
+
+tools/box:
+	curl -Ls https://github.com/humbug/box/releases/download/3.7.0/box.phar -o tools/box && chmod +x tools/box
```
Comment (theofidry) [inline Makefile]: for those targets I would add `touch $@` to avoid the target to get outdated. You sometimes have cases where the target timestamp is not updated properly otherwise making the whole Makefile usage more flimsy. 

This comment applies to the other Makefile targets

File: box.json.dist [hunk 6]
```diff
@@ -0,0 +1,46 @@
+{
+    "base-path": "build/phar",
+    "output": "../pdepend.phar",
+    "compression": "GZ",
+    "directories": ["src/bin", "src/main", "vendor"],
```
Comment (theofidry) [inline box.json.dist]: note that this could be improve (but I would suggest to do that in another PR)

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+.PHONY: phpunit
+
+package: tools/box
+	@rm -rf build/phar && mkdir -p build/phar build/phar/src
+
+	php scripts/update-version.php
+	cp -r src/bin src/main build/phar/src
+	cp -r composer.json LICENSE build/phar
+
+	cd build/phar && composer config platform.php 7.1 && composer update --no-dev -o -a
```
Comment (theofidry) [inline Makefile]: I think the platform needs to be a specific version, e.g. `7.1.3`. `-o -a` are overkill since `-a` implies `-o` and Box will take care of dumping the autoloader in authoritative mode anyway (so doing it here results in the autoloader being dumped twice)
